### DISCUSSION TITLE
[mini.ai]: Paragraph textobject?

### DISCUSSION THREAD

**Discussion by qtyra** [nvim-mini/mini.nvim#1119] [Q&A] [answered]
Related: https://github.com/echasnovski/mini.nvim/discussions/266 and https://github.com/echasnovski/mini.nvim/issues/261

Based on the [docs](https://github.com/echasnovski/mini.ai/blob/main/doc/mini-ai.txt#L294), it seems the best approach is to return a list of paragraph positions. Is there a [neo]vim function to get the selected region of a native text object? The implementation would be quite trivial in that case. I prefer to ask before implementing a buggy paragraph text object and then realize there's an easier way :laughing:

Anyway, i think adding at least the 'p' text object to mini.ai would be a great addition, i miss doing something like 'cinp', and the next best option is '}jcip', which is rather inconvenient. For 'w' and 'W', the alternatives 'wcw', 'WcW' are just as good or even better.

**Comment by echasnovski** [comment, marked as answer]
I am not aware of a better way for paragraph textobject than writing a function which returns either a single target region or an array of regions. I think the latter is easier to implement.

Here is a partial solution showcasing basic approach:

```lua
local inner_paragraph = function()
  local lines = vim.api.nvim_buf_get_lines(0, 0, -1, false)
  local res, cur_paragraph = {}, { vis_mode = 'V' }
  for i, l in ipairs(lines) do
    if l:find('^%s*$') ~= nil then
      if cur_paragraph.from ~= nil then
        cur_paragraph.to = { line = i - 1, col = 1 }
        table.insert(res, cur_paragraph)
        cur_paragraph = { vis_mode = 'V' }
      end
    else
      if cur_paragraph.from == nil then cur_paragraph.from = { line = i, col = 1 } end
    end
  end

  if cur_paragraph.from ~= nil then
    cur_paragraph.to = { line = #lines, col = 1 }
    table.insert(res, cur_paragraph)
  end

  return res
end

local outer_paragraph = function()
  -- Left as homework
end

local paragraph = function(ai_type, _, _)
  return ai_type == 'i' and inner_paragraph() or outer_paragraph()
end

require('mini.ai').setup({
  -- Is needed otherwise paragraphs more than `n_lines` lines will not be selected
  n_lines = 9999,
  custom_textobjects = {
    p = paragraph,
  },
})
```

Adding built-in paragraph textobject is an interesting idea and even probably doable in as `MiniExtra.gen_ai_spec.paragraph`, but the (deliberate) restriction of `config.n_lines` makes it somewhat not usable. So let's leave it as it (at least for now).

------

**Note**: make sure to update to latest `main`, as it has the fix to respect `vis_mode` when output is an array of regions.
**Note 2**: there is a somewhat funny coincidence that almost exactly two years ago (2022-08-05) 'mini.ai' was updated to allow callable with array of regions as output. Life is strange sometimes.

**Comment by qtyra** [comment]
Thanks @echasnovski , i did my homework and implemented outer_paragraph :), in case it helps anyone:

```lua
local outer_paragraph = function()
    local lines = vim.api.nvim_buf_get_lines(0, 0, -1, false)
    local res, cur_paragraph = {}, { vis_mode = "V" }
    for i, l in ipairs(lines) do
        if not l:find("^%s*$") and cur_paragraph.to then
            table.insert(res, cur_paragraph)
            cur_paragraph = { vis_mode = "V" }
        end
        if not l:find("^%s*$") and not cur_paragraph.from and not cur_paragraph.to then
            cur_paragraph.from = { line = i, col = 1 }
        elseif l:find("^%s*$") then
            cur_paragraph.to = { line = i, col = 1 }
        end
    end
    local last_from = cur_paragraph.from.line - 1 > 0 and cur_paragraph.from.line - 1 or 1
    cur_paragraph.from = { line = last_from, col = 1 }
    cur_paragraph.to = { line = #lines, col = 1 }
    table.insert(res, cur_paragraph)
    return res
end
```

**Reply by echasnovski** [reply]
Nice! A- :)
One reason why I didn't want to spend time on this is that handling all edge cases gets tedious. Here are cases in which this will not work as well as built-in `ap`:
- Built-in `ap` prefers including all blank lines after paragraph (this function does also), but for last paragraph which does not have any it falls back on preferring all blank lines before it. This is not the case here.
- It throws error if buffer contains only blank lines. That said, the `inner_paragraph` also doesn't behave consistently (built-in `ip` selects whole buffer, this implementation does nothing), which is another argument as to why implementing this gets tedious.

**Reply by qtyra** [reply]
@echasnovski I've run into a problem with the way mini.ai chooses the best match. Consider this file (with the empty lines around the text):

```

abc

```

and this function

```lua
local outer_paragraph = function()
    return {
        { from = { line = 1, col = 1 }, to = { line = 2, col = 1 }, vis_mode = "V" },
        { from = { line = 2, col = 1 }, to = { line = 3, col = 1 }, vis_mode = "V" },
    }
end
```

It behaves like Vim's 'ap' except when the cursor is exactly at the start of 'abc', it selects the region before the paragraph (check it out with visual mode). Initially, I thought, "Well, since both are the best match following `search_method = cover`, then mini.ai picks the first one". However, this behavior is inconsistent when the cursor is on 'b' or 'c' where selects the empty line *after* the paragraph. I've updated to the latest version, so i don't know if it's a `vis_mode` thing.

Is this behavior expected or it's a bug? If it's expected, I'd greatly appreciate to know if there's a way (even a hacky one) to control it.

**Reply by echasnovski** [reply]
> It behaves like Vim's 'ap' except when the cursor is exactly at the start of 'abc', it selects the region before the paragraph (check it out with visual mode). Initially, I thought, "Well, since both are the best match following `search_method = cover`, then mini.ai picks the first one". However, this behavior is inconsistent when the cursor is on 'b' or 'c' where selects the empty line _after_ the paragraph. I've updated to the latest version, so i don't know if it's a `vis_mode` thing.
> 
> Is this behavior expected or it's a bug? If it's expected, I'd greatly appreciate to know if there's a way (even a hacky one) to control it.

It is somewhat expected. `vis_mode` only controls which visual mode will be used to select textobject, it has nothing to do with computation of best region match.

When cursor is on "b" and onward, it selects second one because it is the only region that covers current single cell selection.
When cursor is on "a", both regions are covering and so happens that have same actual width, so the first one is chosen. If lines are 'xxxxxx', 'abcb', '', then the second one is preferred as *it* has the smaller width. Preferring tighter match in case of several covering regions is [documented](https://github.com/echasnovski/mini.nvim/blob/d8faae693d266ee4f3a2bf1fb7ca4bc81314c3e9/doc/mini-ai.txt#L382-L383).

The best you can do is to properly compute `to.col` to be an actual last column of the line. In this case any cursor position on the line will result into overlapping covering regions and "the smallest width wins" will be consistently applied.

**Reply by qtyra** [reply]
> It is somewhat expected. vis_mode only controls which visual mode will be used to select textobject, it has nothing to do with computation of best region match.

Ok I wrongly assumed that `vis_mode = 'V'` would make mini.ai ignore `col`, it made sense in my mind haha.

> The best you can do is to properly compute to.col to be an actual last column of the line. In this case any cursor position on the line will result into overlapping covering regions and "the smallest width wins" will be consistently applied.

This solution consistently selects the region *before* the paragraph at any position of 'abc', but that's different from the vim's 'ip' behavior which selects the empty lines *after* the paragraph when the cursor is over the text.

I briefly experimented with vim's 'ap' in empty lines between paragraphs and i think i'll stick to the default p as this is getting complicated.

Btw, unrelated but one of the examples in the [docs](https://github.com/echasnovski/mini.nvim/blob/main/doc/mini-ai.txt#L341) works with lowercase snake_case but not with the uppercase version. I managed to make it work for both, but since it only says '... with camel case support.', I'm not sure whether to put it in [show and tell](https://github.com/echasnovski/mini.nvim/discussions/categories/show-and-tell) or make a pull request.

**Reply by echasnovski** [reply]
> Btw, unrelated but one of the examples in the [docs](https://github.com/echasnovski/mini.nvim/blob/main/doc/mini-ai.txt#L341) works with lowercase snake_case but not with the uppercase version. I managed to make it work for both, but since it only says '... with camel case support.', I'm not sure whether to put it in [show and tell](https://github.com/echasnovski/mini.nvim/discussions/categories/show-and-tell) or make a pull request.

No need to change the docs, as it *is* an example for camel case support, not snake case.
